### PR TITLE
ignore HorizontalPodAutoscalerConditionsAnnotation unmarshalling errors

### LINKED ISSUE #88738: Missing validation for HPA annotation
Recently, I tried to create an HPA object with an empty value for the `autoscaling.alpha.kubernetes.io/conditions` annotation and even though the response from apiserver is a 500, the object itself still got created (I verified this by querying for the key from the etcd). And since then `kubectl get/list hpa` calls are failing because apiserver seems to be unable to parse that particular object. Even worse, I wasn't even able to delete that object through the k8s api (had to delete the key directly from etcd).

This was on a v1.14.9 cluster.

Here's a minimal hpa object to reproduce the issue (both `v2beta1` and `v2beta2` seem to have it):
```
{
    "kind": "HorizontalPodAutoscaler",
    "apiVersion": "autoscaling/v2beta1",
    "metadata": {
        "name": "bad-hpa-object",
        "annotations": {
            "autoscaling.alpha.kubernetes.io/conditions": ""
        }
    },
    "spec": {
        "scaleTargetRef": {
            "kind": "Deployment",
            "name": "test",
            "apiVersion": "apps/v1"
        },
        "maxReplicas": 1
    }
}
```

Failing "create hpa" request:
```
$ kubectl create -f bad-hpa-object.json --v=8
...
I0302 18:45:37.912801   25603 request.go:942] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"unexpected end of JSON input","code":500}
I0302 18:45:37.913015   25603 helpers.go:196] server response object: [{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "error when creating \"hpa.json\": unexpected end of JSON input",
  "code": 500
}]
```

Kubectl call failing after above request:
```
$ kubectl get hpa
Error from server: unexpected end of JSON input
```

But object itself is present in etcd and I manually deleted it:
```
$ ETCDCTL_API=3 etcdctl get --keys-only --prefix "/registry/horizontalpodautoscalers/"
/registry/horizontalpodautoscalers/default/bad-hpa-object
$
$ etcdctl del /registry/horizontalpodautoscalers/default/bad-hpa-object
1
```

Which fixed the issue:
```
$ kubectl get hpa
No resources found.
```

cc @kubernetes/sig-api-machinery-bugs (I'm tagging apimachinery here because even though this could be a missing validation for a particular API owned by @kubernetes/sig-autoscaling-bugs , I'm wondering why apiserver created the object even though it returned a 500 to the client)

### PR DESCRIPTION
**Pull request by hakuna-matatah**
Ignore HorizontalPodAutoscalerConditionsAnnotation unmarshalling errors from v1-> internal  because we persist invalid annotations into v1 etcd objects

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR  intentionally ignores validation of HorizontalPodAutoscalerConditionsAnnotation.
We need this because, today we can create HPA objects like it was mentioned in the issue https://github.com/kubernetes/kubernetes/issues/88738, with 
` "autoscaling.alpha.kubernetes.io/conditions": ""`  and it will convert to v1 object  and persist it in etcd without validating the value of the annotation. This triggers a bug when we try to query this object from APIServer/etcd as API Server tries to validate annotations when serving the response.
By ignoring the validation, it will be able to serve response back fine.

**Which issue(s) this PR fixes**:

Fixes #88738 

**Special notes for your reviewer**:

Testing: Ignore unmarshalled annotation errors and successful roundtrip will ensure conversions are happening fine.

**Does this PR introduce a user-facing change?**:
NONE
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

### CODE REVIEW

File: pkg/apis/autoscaling/v1/defaults_test.go [hunk 1]
```diff
@@ -67,6 +70,53 @@ func TestSetDefaultHPA(t *testing.T) {
...
+	}{
+		{
+			hpa: autoscalingv1.HorizontalPodAutoscaler{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						autoscaling.HorizontalPodAutoscalerConditionsAnnotation: "",
+					},
+				},
+			},
+			test: "test empty value for HorizontalPodAutoscalerConditionsAnnotation ",
```
Comment (shyamjvs) [inline pkg/apis/autoscaling/v1/defaults_test.go]: nit: Extra space at the end of string

File: pkg/apis/autoscaling/v1/conversion.go [hunk 2]
```diff
@@ -404,7 +404,10 @@ func Convert_v1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler(i
 	if currentConditionsEnc, hasCurrentConditions := out.Annotations[autoscaling.HorizontalPodAutoscalerConditionsAnnotation]; hasCurrentConditions {
 		var currentConditions []autoscalingv1.HorizontalPodAutoscalerCondition
 		if err := json.Unmarshal([]byte(currentConditionsEnc), &currentConditions); err != nil {
-			return err
+			//ignore unmarshalling errors because we persist invalid annotations into v1 objects
+			// Issue - https://github.com/kubernetes/kubernetes/issues/88738
+			delete(out.Annotations, autoscaling.HorizontalPodAutoscalerConditionsAnnotation)
```
Comment (shyamjvs) [inline pkg/apis/autoscaling/v1/conversion.go]: Not sure if this logic is ok. Could some of the content already be partially written to `currentConditions` above?

@liggitt Could you also take a look if the logic is fine?

**Comment by liggitt** [comment]
tracking for the milestone, this would be a candidate for backporting as well

**Comment by shyamjvs** [comment]
Following presubmit failures seem unrelated:

- pull-kubernetes-e2e-kind failing with some sig-network e2e tests (seem completely unrelated to HPA)
- pull-kubernetes-integration failing with some cascading deletion integration test

But `pull-kubernetes-verify` is failing with:

```
diff -Naupr -x _output /home/prow/go/src/k8s.io/kubernetes/pkg/apis/autoscaling/v1/BUILD /tmp/verify-bazel.sUmRy7/go/src/k8s.io/kubernetes/pkg/apis/autoscaling/v1/BUILD
--- /home/prow/go/src/k8s.io/kubernetes/pkg/apis/autoscaling/v1/BUILD	2020-03-17 03:18:11.298570877 +0000
+++ /tmp/verify-bazel.sUmRy7/go/src/k8s.io/kubernetes/pkg/apis/autoscaling/v1/BUILD	2020-03-17 03:18:44.856646407 +0000
@@ -30,9 +30,11 @@ go_test(
     embed = [":go_default_library"],
     deps = [
         "//pkg/api/legacyscheme:go_default_library",
+        "//pkg/apis/autoscaling:go_default_library",
         "//pkg/apis/autoscaling/install:go_default_library",
         "//pkg/apis/core/install:go_default_library",
         "//staging/src/k8s.io/api/autoscaling/v1:go_default_library",
+        "//staging/src/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_library",
         "//staging/src/k8s.io/apimachinery/pkg/runtime:go_default_library",
         "//vendor/k8s.io/utils/pointer:go_default_library",
     ],

Run ./hack/update-bazel.sh
```

Seems like indenting is off for newly added imports.
@hakuna-matatah Can you fix by running that script?

File: pkg/apis/autoscaling/v1/conversion.go (hunk 2 above)
Comment (hakuna-matatah) [reply pkg/apis/autoscaling/v1/conversion.go]: as long as we are not writing conditions back to `out.Status.Conditions` if/when there is unmarshalling errors, shouldn't that be enough ?

**Comment by liggitt** [comment]
I think we need a more holistic fix to make *sure* we control all serialized annotation data, and that any existing bad persisted annotations are ignored.

I added a commit to https://github.com/liggitt/kubernetes/commits/hpa-conversion that does the following:
* Added a helper method that removes all the annotations used to round-trip HPA fields to old API versions
* Added a helper method to deep copy and return a string map that can be mutated
* Always drop round-trip annotations when going from internal to external
* Only use data from annotations when it unmarshals without errors

The last commit in that branch should resolve this issue. Can you pull in that commit and add a commit on top that unit tests the following scenarios for every external version (v1, v2beta1, v2beta2):
* Start with an external object, adds invalid JSON into all the annotations
* Converts to internal
* Converts back to external
* Verifies the original input object was not modified
* Verifies the final output object only has valid annotations

**Comment by hakuna-matatah** [comment]
> Following presubmit failures seem unrelated:
> 
> * pull-kubernetes-e2e-kind failing with some sig-network e2e tests (seem completely unrelated to HPA)
> * pull-kubernetes-integration failing with some cascading deletion integration test
> 
> But `pull-kubernetes-verify` is failing with:
> 
> ```
> diff -Naupr -x _output /home/prow/go/src/k8s.io/kubernetes/pkg/apis/autoscaling/v1/BUILD /tmp/verify-bazel.sUmRy7/go/src/k8s.io/kubernetes/pkg/apis/autoscaling/v1/BUILD
> --- /home/prow/go/src/k8s.io/kubernetes/pkg/apis/autoscaling/v1/BUILD	2020-03-17 03:18:11.298570877 +0000
> +++ /tmp/verify-bazel.sUmRy7/go/src/k8s.io/kubernetes/pkg/apis/autoscaling/v1/BUILD	2020-03-17 03:18:44.856646407 +0000
> @@ -30,9 +30,11 @@ go_test(
>      embed = [":go_default_library"],
>      deps = [
>          "//pkg/api/legacyscheme:go_default_library",
> +        "//pkg/apis/autoscaling:go_default_library",
>          "//pkg/apis/autoscaling/install:go_default_library",
>          "//pkg/apis/core/install:go_default_library",
>          "//staging/src/k8s.io/api/autoscaling/v1:go_default_library",
> +        "//staging/src/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_library",
>          "//staging/src/k8s.io/apimachinery/pkg/runtime:go_default_library",
>          "//vendor/k8s.io/utils/pointer:go_default_library",
>      ],
> 
> Run ./hack/update-bazel.sh
> ```
> 
> Seems like indenting is off for newly added imports.
> @hakuna-matatah Can you fix by running that script?

Fixed.

**Comment by hakuna-matatah** [comment]
Oops!! accidentally closed the PR while responding and resolving Shyam's comment.

**Comment by hakuna-matatah** [comment]
> I think we need a more holistic fix to make _sure_ we control all serialized annotation data, and that any existing bad persisted annotations are ignored.
> 
> I added a commit to https://github.com/liggitt/kubernetes/commits/hpa-conversion that does the following:
> 
> * Added a helper method that removes all the annotations used to round-trip HPA fields to old API versions
> * Added a helper method to deep copy and return a string map that can be mutated
> * Always drop round-trip annotations when going from internal to all external versions, prior to setting any annotations to hold round-tripped data
> * Only use data from annotations when it unmarshals without errors
> 
> The last commit in that branch should resolve this issue. Can you pull in that commit and add a commit on top that unit tests the following scenarios for every external version (v1, v2beta1, v2beta2):
> 
> * Start with an external object, adds invalid JSON into all the annotations
> * Converts to internal
> * Converts back to external
> * Verifies the original input object was not modified (can deepcopy the input object prior to starting the test, then verify it matches the input object after conversion using reflect.DeepEqual())
> * Verifies the final output object only has valid annotations

Sure @liggitt. I will pull in that commit and update this PR with tests you mentioned.

**Comment by smourapina** [comment]
Hello @hakuna-matatah! 
I am a member of Bug Triage for 1.18, with a reminder that you will need to propose a cherry-pick for this PR. You can find information on the process here: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md

**Comment by smourapina** [comment]
One more thing: cherry pick deadline for 1.18 is today at EOD PST. Would be good if you could make it by then @hakuna-matatah. Thanks!

**Comment by hakuna-matatah** [comment]
> One more thing: cherry pick deadline for 1.18 is today at EOD PST. Would be good if you could make it by then @hakuna-matatah. Thanks!

Oh! I didn't know that. Sorry.

> One more thing: cherry pick deadline for 1.18 is today at EOD PST. Would be good if you could make it by then @hakuna-matatah. Thanks!

Woah. I didn't know that. Will rollout a fix now.

**Comment by markyjackson-taulia** [comment]
I wanted to check back in on this. It doesn’t seem we are going to make v1.18 and may have to aim for v1.18.1.

**Comment by liggitt** [comment]
pulled in the test commit and fixed up round-trip test failures in #89963

/close

**Pull request closed without merging**